File: common/src/main/java/juuxel/adorn/block/BrewerBlock.java
package juuxel.adorn.block;

import com.mojang.serialization.MapCodec;
import juuxel.adorn.block.entity.BrewerBlockEntity;
import juuxel.adorn.lib.AdornStats;
import juuxel.adorn.util.ShapeRotation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.BaseEntityBlock;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.Containers;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.util.RandomSource;
import net.minecraft.world.phys.shapes.VoxelShape;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.Nullable;

import java.util.Map;

public final class BrewerBlock extends VisibleBlockWithEntity implements BlockWithDescription {
    public static final EnumProperty<Direction> FACING = BlockStateProperties.HORIZONTAL_FACING;
    public static final BooleanProperty HAS_MUG = BooleanProperty.create("has_mug");
    public static final BooleanProperty ACTIVE = BooleanProperty.create("active");

    private static final Map<Direction, VoxelShape> SHAPES = ShapeRotation.mergeShapeMaps(
        ShapeRotation.buildShapeRotationsFromNorth(4, 0, 2, 12, 2, 12),
        ShapeRotation.buildShapeRotationsFromNorth(4, 2, 8, 12, 8, 12),
        ShapeRotation.buildShapeRotationsFromNorth(4, 8, 2, 12, 14, 12)
    );
    private static final double RANDOM_CLOUD_OFFSET = 0.0625;
    private static final double FACING_CLOUD_OFFSET = 0.2;

    public BrewerBlock(Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState().setValue(HAS_MUG, false).setValue(ACTIVE, false));
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return defaultBlockState().setValue(FACING, ctx.getHorizontalDirection().getOpposite());
    }

    @Override
    protected InteractionResult useWithoutItem(BlockState state, Level world, BlockPos pos, Player player, BlockHitResult hit) {
        if (world.isClientSide()) return InteractionResult.SUCCESS;

        if (world.getBlockEntity(pos) instanceof BrewerBlockEntity brewer) {
            player.openMenu(brewer);
            player.awardStat(AdornStats.OPEN_BREWER);
        }

        return InteractionResult.CONSUME;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return SHAPES.get(state.getValue(FACING));
    }

    @Override
    protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel world, BlockPos pos, boolean moved) {
        Containers.updateNeighboursAfterDestroy(state, world, pos);
    }

    @Override
    public @Nullable BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return AdornBlockEntities.BREWER.get().create(pos, state);
    }

    @Override
    public <T extends BlockEntity> @Nullable BlockEntityTicker<T> getTicker(Level world, BlockState stateUnused, BlockEntityType<T> type) {
        return world instanceof ServerLevel serverWorld ? createTickerHelper(
            type,
            AdornBlockEntities.BREWER.get(),
            (worldUnused, pos, state, blockEntity) -> BrewerBlockEntity.tick(serverWorld, pos, state, blockEntity)
        ) : null;
    }

    @Override
    public BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
    }

    @Override
    public BlockState mirror(BlockState state, Mirror mirror) {
        return state.rotate(mirror.getRotation(state.getValue(FACING)));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(FACING, HAS_MUG, ACTIVE);
    }

    @Override
    public boolean hasAnalogOutputSignal(BlockState state) {
        return true;
    }

    @Override
    protected int getAnalogOutputSignal(BlockState state, Level world, BlockPos pos, Direction direction) {
        return world.getBlockEntity(pos) instanceof BrewerBlockEntity brewer ? brewer.calculateComparatorOutput() : 0;
    }

    @Override
    public void animateTick(BlockState state, Level world, BlockPos pos, RandomSource random) {
        if (state.getValue(ACTIVE) && random.nextInt(3) == 0) {
            var facing = state.getValue(FACING);
            double x = pos.getX() + 0.5 + random.nextDouble() * RANDOM_CLOUD_OFFSET + facing.getStepX() * FACING_CLOUD_OFFSET;
            double y = pos.getY() + 0.37 + random.nextDouble() * RANDOM_CLOUD_OFFSET;
            double z = pos.getZ() + 0.5 + random.nextDouble() * RANDOM_CLOUD_OFFSET + facing.getStepZ() * FACING_CLOUD_OFFSET;
            world.addParticle(ParticleTypes.CLOUD, x, y, z, 0.0, 0.0, 0.0);
        }
    }

    @Override
    protected MapCodec<? extends BaseEntityBlock> codec() {
        throw new UnsupportedOperationException();
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/BrewerBlock.kt
package juuxel.adorn.block

import juuxel.adorn.block.entity.BrewerBlockEntity
import juuxel.adorn.lib.AdornStats
import juuxel.adorn.util.buildShapeRotationsFromNorth
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityTicker
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.block.entity.LootableContainerBlockEntity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.particle.ParticleTypes
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.DirectionProperty
import net.minecraft.state.property.Properties
import net.minecraft.util.ActionResult
import net.minecraft.util.BlockMirror
import net.minecraft.util.BlockRotation
import net.minecraft.util.Hand
import net.minecraft.util.ItemScatterer
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.BlockView
import net.minecraft.world.World
import java.util.Random

class BrewerBlock(settings: Settings) : VisibleBlockWithEntity(settings) {
    init {
        defaultState = defaultState.with(HAS_MUG, false).with(ACTIVE, false)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState =
        defaultState.with(FACING, ctx.playerFacing.opposite)

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        if (world.isClient) return ActionResult.SUCCESS

        val blockEntity = world.getBlockEntity(pos)
        if (blockEntity is BrewerBlockEntity) {
            player.openMenu(blockEntity)
            player.incrementStat(AdornStats.OPEN_BREWER)
        }

        return ActionResult.CONSUME
    }

    override fun getOutlineShape(state: BlockState, world: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape =
        SHAPES[state[FACING]]!!

    override fun onPlaced(world: World, pos: BlockPos, state: BlockState, entity: LivingEntity?, stack: ItemStack) {
        if (stack.hasCustomName()) {
            (world.getBlockEntity(pos) as? LootableContainerBlockEntity)?.customName = stack.name
        }
    }

    override fun onStateReplaced(state: BlockState, world: World, pos: BlockPos, newState: BlockState, moved: Boolean) {
        if (!state.isOf(newState.block)) {
            val entity = world.getBlockEntity(pos)

            if (entity is Inventory) {
                ItemScatterer.spawn(world, pos, entity)
                world.updateComparators(pos, this)
            }

            super.onStateReplaced(state, world, pos, newState, moved)
        }
    }

    override fun createBlockEntity(pos: BlockPos, state: BlockState): BlockEntity? =
        AdornBlockEntities.BREWER.instantiate(pos, state)

    override fun <T : BlockEntity> getTicker(world: World, state: BlockState, type: BlockEntityType<T>): BlockEntityTicker<T>? =
        if (!world.isClient) checkType(type, AdornBlockEntities.BREWER, BrewerBlockEntity::tick) else null

    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState =
        state.with(FACING, rotation.rotate(state[FACING]))

    override fun mirror(state: BlockState, mirror: BlockMirror): BlockState =
        state.rotate(mirror.getRotation(state[FACING]))

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(FACING, HAS_MUG, ACTIVE)
    }

    override fun hasComparatorOutput(state: BlockState): Boolean = true

    override fun getComparatorOutput(state: BlockState, world: World, pos: BlockPos): Int =
        (world.getBlockEntity(pos) as? BrewerBlockEntity)?.calculateComparatorOutput() ?: 0

    override fun randomDisplayTick(state: BlockState, world: World, pos: BlockPos, random: Random) {
        if (state[ACTIVE] && random.nextInt(3) == 0) {
            val facing = state[FACING]
            val x = pos.x + 0.5 + random.nextDouble() * RANDOM_CLOUD_OFFSET + facing.offsetX * FACING_CLOUD_OFFSET
            val y = pos.y + 0.37 + random.nextDouble() * RANDOM_CLOUD_OFFSET
            val z = pos.z + 0.5 + random.nextDouble() * RANDOM_CLOUD_OFFSET + facing.offsetZ * FACING_CLOUD_OFFSET
            world.addParticle(ParticleTypes.CLOUD, x, y, z, 0.0, 0.0, 0.0)
        }
    }

    companion object {
        val FACING: DirectionProperty = Properties.HORIZONTAL_FACING
        val HAS_MUG: BooleanProperty = BooleanProperty.of("has_mug")
        val ACTIVE: BooleanProperty = BooleanProperty.of("active")
        private val SHAPES = buildShapeRotationsFromNorth(4, 0, 2, 12, 14, 12)
        private const val RANDOM_CLOUD_OFFSET = 0.0625
        private const val FACING_CLOUD_OFFSET = 0.2
    }
}
